certbot/cli.py

```diff
@@ -848,29 +848,33 @@ def prepare_and_parse_args(plugins, args, detect_defaults=False):  # pylint: dis
         "e.g. -vvv.")
     helpful.add(
         None, "-t", "--text", dest="text_mode", action="store_true",
-        help=argparse.SUPPRESS)
+        default=flag_default("text_mode"), help=argparse.SUPPRESS)
     helpful.add(
-        None, "--max-log-backups", type=nonnegative_int, default=1000,
+        None, "--max-log-backups", type=nonnegative_int,
+        default=flag_default("max_log_backups"),
         help="Specifies the maximum number of backup logs that should "
              "be kept by Certbot's built in log rotation. Setting this "
              "flag to 0 disables log rotation entirely, causing "
              "Certbot to always append to the same log file.")
     helpful.add(
         [None, "automation", "run", "certonly"], "-n", "--non-interactive", "--noninteractive",
         dest="noninteractive_mode", action="store_true",
+        default=flag_default("noninteractive_mode"),
         help="Run without ever asking for user input. This may require "
               "additional command line flags; the client will try to explain "
               "which ones are required if it finds one missing")
     helpful.add(
         [None, "register", "run", "certonly"],
         constants.FORCE_INTERACTIVE_FLAG, action="store_true",
+        default=flag_default("force_interactive"),
         help="Force Certbot to be interactive even if it detects it's not "
              "being run in a terminal. This flag cannot be used with the "
              "renew subcommand.")
     helpful.add(
         [None, "run", "certonly", "certificates"],
         "-d", "--domains", "--domain", dest="domains",
-        metavar="DOMAIN", action=_DomainsAction, default=[],
+        metavar="DOMAIN", action=_DomainsAction,
+        default=flag_default('domains')[:], # Copy, not reference
```

Comment:
I'm slightly worried about other mutable defaults such as `pref_challs`. What if we made `flag_default` return a copy of the object using the `copy` module?

Reply:
That makes sense. As I was incorporating your feedback, I found that `ParseTest.test_cli_ini_domains` is causing the modification to CLI_DEFAULTS. It uses a mocked `flag_default`, so I'm having the mock return a copy as well.

<sleepy braindump>
In the commit you just reviewed, `test_cli_ini_domains` mocked `flag_default` in a way that didn't return a copy, and I just verified with pdb that the value in `CLI_DEFAULTS` does change after parsing domains from the temp file. When I modified `parse_default` to return a copy, given your feedback, the corrupted list was returned.

The default given to the argparser as you just reviewed was a copy of `domains`, as shown in the snippet above. Shouldn't that have likewise returned a copy of a corrupted list, containing artifacts from `test_cli_ini_domains`? Whack.
</braindump>

Reply:
Hmm. Happy to help here but I'm not sure I understand.

If I checkout the commit I just reviewed (955ebb8) and break after the calls to `self.parse` after writing the temporary file, `constants.CLI_DEFAULTS['domains']` is still an empty list.

Reply:
There are two cases I'm interested in. In both cases, `test_cli_ini_domains` does what we'd consider the "wrong" thing (not copying the retrieval). In one case, `cli.flag_defaults` returns a copy of the retrieval. In the second case, `flag_defaults` returns a reference which is only copied when it's provided as an argument to `add`. The second case works and the first doesn't when I'd have thought they were equivalent.

A branch demonstrating this in two commits is at https://github.com/cj-dev/certbot/compare/cj-dev:4435-cli-defaults...4435-but-broken. Commit [2e30494](`https://github.com/cj-dev/certbot/commit/2e3049417d7a9a13cd3f8c19414baa19d14626b9`) works, commit [b9df87c](https://github.com/cj-dev/certbot/commit/b9df87ccfb78f9af9455377fd22657b53944f252) doesn't. Only `python certbot/tests/cli_test.py ParseTest.test_cli_ini_domains` is necessary to see the effect.